cache/refs.go

```diff
@@ -1517,3 +1523,94 @@ func readonlyOverlay(opt []string) []string {
 	}
 	return out
 }
+
+func setSharable(mounts snapshot.Mountable) snapshot.Mountable {
+	return &sharableMountable{Mountable: mounts}
+}
+
+// sharableMountable allows sharing underlying (possibly writable) mounts among callers.
+// This is useful to share writable overlayfs mounts.
+//
+// NOTE: Mount() method doesn't return the underlying mount configuration (e.g. overlayfs mounts)
+//       instead it always return bind mounts of the temporary mount point. So if the caller
+//       needs to inspect the underlying mount configuration (e.g. for optimized differ for
+//       overlayfs), this wrapper shouldn't be used.
+type sharableMountable struct {
+	snapshot.Mountable
+
+	count int32
+	mu    sync.Mutex
+
+	curMounts     []mount.Mount
+	curMountPoint string
+	curRelease    func() error
+}
+
+func (sm *sharableMountable) Mount() ([]mount.Mount, func() error, error) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	if sm.curMounts == nil {
+		mounts, release, err := sm.Mountable.Mount()
+		if err != nil {
+			return nil, nil, err
+		}
+		var isOverlay bool
+		for _, m := range mounts {
+			if m.Type == "overlay" {
+				isOverlay = true
+			}
+		}
+		if !isOverlay {
+			// Don't need temporary mount wrapper for non-overlayfs mounts
+			return mounts, release, nil
+		}
+		dir, err := ioutil.TempDir("", "buildkit")
+		if err != nil {
+			return nil, nil, err
+		}
+		if err := mount.All(mounts, dir); err != nil {
+			return nil, nil, err
+		}
+		sm.curMounts = []mount.Mount{
+			{
+				Source: dir,
+				Type:   "bind",
+				Options: []string{
+					"rw",
+					"rbind",
+				},
+			},
+		}
+		sm.curMountPoint = dir
+		sm.curRelease = release
+	}
+
+	mounts := make([]mount.Mount, len(sm.curMounts))
+	copy(mounts, sm.curMounts)
+
+	sm.count++
+	return mounts, func() error {
+		sm.mu.Lock()
+		defer sm.mu.Unlock()
+
+		sm.count--
+		if sm.count < 0 {
+			if v := os.Getenv("BUILDKIT_DEBUG_PANIC_ON_ERROR"); v == "1" {
+				panic("release of released mount")
+			}
+		} else if sm.count > 0 {
+			return nil
+		}
+
+		// no mount exist. release the current mount.
+		sm.curMounts = nil
+		if err := mount.Unmount(sm.curMountPoint, 0); err != nil {
+			return err
+		}
```

Comment:
This doesn't have to be fixed in this PR as it's a larger issue, but if buildkitd crashed or errors were encountered unmounting here I guess we could leak overlay mounts. Then if buildkitd started up again, you could run into the same issue of duplicating overlay mounts because the old ones are still sitting around. This is different from mounts made for containers w/ runc/containerd because those are typically mounted inside a mount namespace, meaning they will get auto cleaned up by the kernel when no processes are left in the namespace. The only remediation would be to manually unmount stuff from under /tmp or reboot the host.

Like I said, this is probably a pre-existing issue, but something to keep in mind. We might want a way of cleaning up any leftover mounts when buildkitd starts up (though even that can't be 100% robust if it's possible for users to change the configuration of where such tmp mounts are made). Maybe we could even run buildkitd inside its own mount namespace, but trying to get Go to do that is never easy and might have weird undesirable side effects.

Reply:
Thank you for the comment. Let's work on it in following-up PRs. I think we can add a directory under `/var/lib/buildkit/` for storing all temporary overlay mounts. When buildkitd starts, it should clean up all existing mounts under that directory.